src/systems/navsat/NavSat.cc

```diff
@@ -0,0 +1,287 @@
+  /// `PostUpdate`, so that components can be created during the next
+  /// `PreUpdate`.
+  public: std::unordered_set<Entity> newSensors;
+
+  /// When the system is just loaded, we loop over all entities to create
+  /// sensors. After this initialization, we only check inserted entities.
+  public: bool initialized = false;
+
+  /// \brief Create sensors in ign-sensors
+  /// \param[in] _ecm Immutable reference to ECM.
+  public: void CreateSensors(const EntityComponentManager &_ecm);
+
+  /// \brief Update sensor data based on physics data
+  /// \param[in] _ecm Immutable reference to ECM.
+  public: void Update(const EntityComponentManager &_ecm);
+
+  /// \brief Remove sensors if their entities have been removed from simulation.
+  /// \param[in] _ecm Immutable reference to ECM.
+  public: void RemoveSensors(const EntityComponentManager &_ecm);
+
+  /// \brief Create sensor
+  /// \param[in] _ecm Immutable reference to ECM.
+  /// \param[in] _entity Sensor entity
+  /// \param[in] _navsat NavSat component.
+  /// \param[in] _parent Parent entity component.
+  public: void AddSensor(
+    const EntityComponentManager &_ecm,
+    const Entity _entity,
+    const components::NavSat *_navSat,
+    const components::ParentEntity *_parent);
+};
+
+//////////////////////////////////////////////////
+NavSat::NavSat() : System(), dataPtr(utils::MakeUniqueImpl<Implementation>())
+{
+}
+
+//////////////////////////////////////////////////
+void NavSat::PreUpdate(const UpdateInfo &/*_info*/,
+    EntityComponentManager &_ecm)
+{
+  IGN_PROFILE("NavSat::PreUpdate");
+
+  // Create components
+  for (auto entity : this->dataPtr->newSensors)
+  {
+    auto it = this->dataPtr->entitySensorMap.find(entity);
+    if (it == this->dataPtr->entitySensorMap.end())
+    {
+      ignerr << "Entity [" << entity
+             << "] isn't in sensor map, this shouldn't happen." << std::endl;
+      continue;
+    }
+    // Set topic
+    _ecm.CreateComponent(entity, components::SensorTopic(it->second->Topic()));
+  }
+  this->dataPtr->newSensors.clear();
+}
+
+//////////////////////////////////////////////////
+void NavSat::PostUpdate(const UpdateInfo &_info,
+                           const EntityComponentManager &_ecm)
+{
+  IGN_PROFILE("NavSat::PostUpdate");
+
+  // \TODO(anyone) Support rewind
+  if (_info.dt < std::chrono::steady_clock::duration::zero())
+  {
+    ignwarn << "Detected jump back in time ["
+        << std::chrono::duration_cast<std::chrono::seconds>(_info.dt).count()
+        << "s]. System may not work properly." << std::endl;
+  }
+
+  this->dataPtr->CreateSensors(_ecm);
+
+  // Only update and publish if not paused.
+  if (!_info.paused)
+  {
+    this->dataPtr->Update(_ecm);
+
+    for (auto &it : this->dataPtr->entitySensorMap)
+    {
+      // Update measurement time
+      auto time = math::durationToSecNsec(_info.simTime);
+      dynamic_cast<sensors::Sensor *>(it.second.get())->Update(
```

Comment:
I don't think the dynamic cast here is necessary. It looks like the `Update` method is overridden by the `NavSatSensor` class: https://github.com/ignitionrobotics/ign-sensors/blob/4015d876d227c98612a1ccde11e88d6178931ab3/include/ignition/sensors/NavSatSensor.hh#L71-L75

Reply:
Nice catch again! We do need to call the base class's update function, but the call can be simplified, see d906b0e.

See the discussion in https://github.com/ignitionrobotics/ign-gazebo/pull/1159 for why this is the case. I'll also take the opportunity to fix all sensors now.

Reply:
https://github.com/ignitionrobotics/ign-gazebo/pull/1282